### PR TITLE
Add source parameter

### LINKED ISSUE #181: `/search?source' parameter enables scoping of search by origin of source data
A user should be able to scope their search by the source of the data.

For instance:

`GET <pelias>/search?text=123 Broadway, NYC&source=openstreetmap`

should only return valid matches from OpenStreetMap, and not OpenAddresses.

This is fundamentally different from Layers, because while Layers may refer to parts of a dataset or even the whole dataset (e.g. OSMWays + OSMnodes), Source only reflects the provenance of that dataset.

### PR DESCRIPTION
**Pull request by orangejulius**
This adds the source parameter described #181. It allows limiting search results to only those matching a particular data source (geonames, quattroshapes, openaddresses, and openstreetmap are the currently available sources).

The filtering done with this parameter is _cumulative_ with the layers parameter, and it is possible to specify combinations of layers and source parameters that can't possibly return any results, such as source=quattroshapes&layers=address. Additionally, the address parser uses some logic to determine when it is safe to query only the admin layer for performance improvement (for example, queries with a text parameter only a few characters long).

For that reason, this branch also contains a major refactoring to all code relating to layers, as well as the address parser. It splits the address parser functionality into two sections, one dealing with the actual address parsing, and one dealing with determining when the search results can be constrained to the admin layer. It also refactors address parsing and layer parameter handling code to simply submit a requested list of types, rather than directly manipulating the list of types sent to Elasticsearch.
Finally, new logic is added in middleware/types and helpers/types to look at types specified by the layers parameter, the source parameter, and the address parser, and decide how to combine them.

The types middleware detects when a combination of layers and source parameter won't return any results, and provides a friendly error message. This allows the full power of layers and source to be used in combination when it makes sense, while ensuring combinations that don't make sense are discoverable quickly.

Fixes #181

### CODE REVIEW

File: controller/search.js [hunk 1]
```diff
@@ -16,13 +15,9 @@ function setup( backend, query ){
...
     };
 
-    if (req.clean.layers) {
-      cmd.type = req.clean.layers;
-    }
-
-    // set type if input suggests targeting a layer(s)
-    if (req.clean.default_layers_set && req.clean.parsed_input) {
-      cmd.type = req.clean.parsed_input.target_layer || cmd.type;
+    if (req.clean.type !== undefined) {
```
Comment (dianashk) [inline controller/search.js]: better to use hasOwnProperty

File: query/sources.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Mapping from data sources to type values
+ */
+
+module.exports = {
+  'geonames': ['geoname'],
+  'openaddresses' : ['openaddresses'],
+  'quattroshapes': ['admin0', 'admin1', 'admin2', 'neighborhood', 'locality', 'local_admin'],
+  'openstreetmap' : ['osmaddress', 'osmnode', 'osmway']
+};
```
Comment (dianashk) [inline query/sources.js]: I like the idea of keeping the mappings for source in a separate file. We should refactor the lookup code I wrote in the response commit to use the external lookup, and this file will house the `alias`->`list` as well as `name`->`alias` mappings. No good reason to have in 2 different places.

**Comment by orangejulius** [comment]
So this is an example response from the query `/v1/search?input=boston&source=geonames`

Note that there is a new section under query, types. This section shows what Elasticsearch types various areas of the code are asking for. So here the address parser is saying "it's ok to search anything", because all the types we have are listed, and the source parameter is saying "please only search for documents with type geoname". There's a types helper that combines all these different preferences, for example the address parser's preferences shouldn't ever override anything set explicitly.

The question is how much of this detail do we want to show? Will end users care about this types section of query, which is really just an intermediate set of data? Or do they want to know the final collection of types that are searched (currently this is hidden but can easily be shown)

```
{
   "geocoding":{
      "version":"0.1",
      "attribution":"http://pelias.mapzen.com/attribution",
      "query":{
         "input":"boston",
         "parsed_input":{
            "regions":[
               "boston"
            ]
         },
         "types":{
            "from_address_parsing":[
               "geoname",
               "osmnode",
               "osmway",
               "admin0",
               "admin1",
               "admin2",
               "neighborhood",
               "locality",
               "local_admin"
            ],
            "from_source":[
               "geoname"
            ]
         },
         "size":10,
         "details":true
      },
      "engine":{
         "name":"Pelias",
         "author":"Mapzen",
         "version":"1.0"
      },
      "timestamp":1441736493479
   },
   "type":"FeatureCollection",
   "features":[

   ]
}
```

**Comment by orangejulius** [comment]
Here's another example of the query `http://localhost:3100/v1/search?input=boston&source=openstreetmap&layers=address`, showing both the intermediate and final types.

```
{
   "geocoding":{
      "version":"0.1",
      "attribution":"http://pelias.mapzen.com/attribution",
      "query":{
         "input":"boston",
         "parsed_input":{
            "regions":[
               "boston"
            ]
         },
         "types":{
            "from_address_parsing":[
               "geoname",
               "osmnode",
               "osmway",
               "admin0",
               "admin1",
               "admin2",
               "neighborhood",
               "locality",
               "local_admin"
            ],
            "from_layers":[
               "osmaddress",
               "openaddresses"
            ],
            "from_source":[
               "osmaddress",
               "osmnode",
               "osmway"
            ]
         },
         "size":10,
         "details":true,
         "type":[
            "osmaddress"
         ]
      },
      "engine":{
         "name":"Pelias",
         "author":"Mapzen",
         "version":"1.0"
      },
      "timestamp":1441737741825
   },
   "type":"FeatureCollection",
   "features":[
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287624245",
               "layer":"address",
               "source":"osm",
               "name":"634 Boston Avenue",
               "housenumber":"634",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"634 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.294812,
               37.924183
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287624628",
               "layer":"address",
               "source":"osm",
               "name":"643 Boston Avenue",
               "housenumber":"643",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"643 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.295115,
               37.924497
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287624665",
               "layer":"address",
               "source":"osm",
               "name":"644 Boston Avenue",
               "housenumber":"644",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"644 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.29499,
               37.924121
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287624866",
               "layer":"address",
               "source":"osm",
               "name":"653 Boston Avenue",
               "housenumber":"653",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"653 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.295307,
               37.924451
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287624888",
               "layer":"address",
               "source":"osm",
               "name":"654 Boston Avenue",
               "housenumber":"654",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"654 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.295201,
               37.924042
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287625027",
               "layer":"address",
               "source":"osm",
               "name":"663 Boston Avenue",
               "housenumber":"663",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"663 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.295525,
               37.924425
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287625074",
               "layer":"address",
               "source":"osm",
               "name":"664 Boston Avenue",
               "housenumber":"664",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"664 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.295435,
               37.924029
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287625203",
               "layer":"address",
               "source":"osm",
               "name":"673 Boston Avenue",
               "housenumber":"673",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"673 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.295716,
               37.924395
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287625212",
               "layer":"address",
               "source":"osm",
               "name":"674 Boston Avenue",
               "housenumber":"674",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"674 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.29566,
               37.923994
            ]
         }
      },
      {
         "type":"Feature",
         "properties":{
            "geocoding":{
               "id":"address-osmway-287625309",
               "layer":"address",
               "source":"osm",
               "name":"683 Boston Avenue",
               "housenumber":"683",
               "street":"Boston Avenue",
               "category":[

               ],
               "country_a":"USA",
               "country":"United States",
               "region":"California",
               "region_a":"CA",
               "county":"San Joaquin County",
               "locality":"Stockton",
               "neighbourhood":"The Homestead",
               "confidence":0.54,
               "label":"683 Boston Avenue, Stockton, CA"
            }
         },
         "geometry":{
            "type":"Point",
            "coordinates":[
               -121.295891,
               37.924346
            ]
         }
      }
   ],
   "bbox":[
      -121.295891,
      37.923994,
      -121.294812,
      37.924497
   ]
}
```

**Comment by dianashk** [comment]
```javascript
{
   "geocoding":{
      "version":"0.1",
      "attribution":"http://pelias.mapzen.com/attribution",
      "query":{
         "input":"boston",
         "parsed_input":{
            "regions":[
               "boston"
            ],
            "target_data_sources":[
               "geoname",
               "osmnode",
               "osmway",
               "admin0",
               "admin1",
               "admin2",
               "neighborhood",
               "locality",
               "local_admin"
            ]
         },
         "layers":[
               "osmaddress",
               "openaddresses"
         ],
         "sources":[
            "osmaddress",
            "osmnode",
            "osmway"
         ],
         "size":10,
         "details":true,
         "type":[
            "osmaddress"
         ]
      },
      "engine":{
         "name":"Pelias",
         "author":"Mapzen",
         "version":"1.0"
      },
      "timestamp":1441737741825
   },
...
```

**Pull request merged**